# Lab 2: Data Types


For this lab, be sure to download the zipped file from Canvas, and after unzipping that, move all those files into the same directory as this Jupyter notebook file.

In this lab, you'll first see how to represent and manipulate another fundamental type of data: text.  A piece of text is called a *string* in Python. You'll also see how to invoke *methods*.  A method is very similar to a function, but tied to a particular piece of data. 

Lastly, you'll begin to see how to work with *collections* of data, like the numbers 2 through 5 or the words "welcome", "to", and "lab".

**Your Name:**

## 1. Review: The building blocks of Python code

The two building blocks of Python code are *expressions* and *statements*.  An **expression** is a piece of code that

* is self-contained, meaning it would make sense to write it on a line by itself, and
* usually has a value.


Here are two expressions that both evaluate to 3

```python
3
5 - 2
```

One important form of an expression is the call expression, which first names a function and then describes its arguments. The function returns some value, based on its arguments. Some important mathematical functions are

| Function | Description                                                   |
|----------|---------------------------------------------------------------|
| `abs`      | Returns the absolute value of its argument                    |
| `max`      | Returns the maximum of all its arguments                      |
| `min`      | Returns the minimum of all its arguments                      |
| `pow`      | Raises its first argument to the power of its second argument |
| `round`    | Round its argument to the nearest integer                     |

Here are two call expressions that both evaluate to 3

```python
abs(2 - 5)
max(round(2.8), min(pow(2, 10), -1 * pow(2, 10)))
```

All these expressions but the first are **compound expressions**, meaning that they are actually combinations of several smaller expressions.  `2 + 3` combines the expressions `2` and `3` by addition.  In this case, `2` and `3` are called **subexpressions** because they're expressions that are part of a larger expression.

A **statement** is a whole line of code.  Some statements are just expressions.  The expressions listed above are examples.

Other statements *make something happen* rather than *having a value*.  After they are run, something in the world has changed.  For example, an assignment statement assigns a value to a name. 

A good way to think about this is that we're **evaluating the right-hand side** of the equals sign and **assigning it to the left-hand side**. Here are some assignment statements:
    
```python
height = 1.3
the_number_five = abs(-5)
absolute_height_difference = abs(height - 1.688)
```

A key idea in programming is that large, interesting things can be built by combining many simple, uninteresting things.  The key to understanding a complicated piece of code is breaking it down into its simple components.

For example, a lot is going on in the last statement above, but it's really just a combination of a few things.  This picture describes what's going on.

<img src="statement.jpg">


---

**Question 1.1:** In the next cell, assign the name `new_year` to the larger number among the following two numbers:

1. the absolute value of $2^{5} - 2^{11} - 2^3$, and 
2. $2 \times 3 \times 337$.

Try to use just one statement (one line of code).

In [51]:
new_year = abs(2**5-2**11-2**3)
new_year

2024

## 2. Text
Text is one of the most common types of values used in programs. A snippet of text is represented by a **string value** in Python. The word "*string*" is a programming term for a sequence of characters. A string might contain a single character, a word, a sentence, or a whole book.

To distinguish text data from actual code, we demarcate strings by putting quotation marks around them. Single quotes (`'`) and double quotes (`"`) are both valid, but the types of opening and closing quotation marks must match. The contents can be any sequence of characters, including numbers and symbols. 

We've seen strings before in `print` statements.  Below, two different strings are passed as arguments to the `print` function.

In [ ]:
print("I <3", 'Data Science')

Note that a space is automatically added between the two (comma-separated) strings passed to `print`.

Just like names can be given to numbers, names can be given to string values.  The names and strings aren't required to be similar in any way. Any name can be assigned to any string.

In [ ]:
one = "two"
plus = "*"
print(one, plus, one)

### 2.1. String Methods

Strings can be transformed using **methods**, which are functions that involve an existing string and some other arguments. One example is the `replace` method, which replaces all instances of some part of a string with some alternative. 

A method is invoked on a string by placing a `.` after the string value, then the name of the method, and finally parentheses containing the arguments. Here's a sketch, where the `<` and `>` symbols aren't part of the syntax; they just mark the boundaries of sub-expressions.

    <expression that evaluates to a string>.<method name>(<argument>, <argument>, ...)

Try to predict the output of these examples, then execute them.

In [8]:
# Replace one letter
"Hello".replace("o", "a")

'Hella'

In [9]:
# Replace a sequence of letters, which appears twice
"hitchhiker".replace("hi", "ma")

'matchmaker'

Once a name is bound to a string value, methods can be invoked on that name as well. The name doesn't change in this case, so a new name is needed to capture the result. 

In [10]:
sharp = "edged"
hot = sharp.replace("ed", "ma")
print(f"sharp: {sharp}")
print(f"hot: {hot}")

sharp: edged
hot: magma


You can call functions on the results of other functions.  For example,

```python
max(abs(-5), abs(3))
```

has value 5.  Similarly, you can invoke methods on the results of other method (or function) calls.

In [11]:
# Calling replace on the output of another call to
# replace
"train".replace("t", "ing").replace("in", "de")

'degrade'

Here's a picture of how Python evaluates a "chained" method call like that:

<img src="chaining_method_calls.jpg"/>

---

**Question 2.1.1:** Redefine the string named `messaging` below by replacing all occurrences of the `Alice` with `Bob` and vice-versa.  (So, your output should be `Bob messaged Alice, but Alice didn't recognize Bob's new number.`)  You *must* use the `replace` method.

*Hint:* When performing many operations to a variable, sometimes it's helpful to see its value after each step.  You can simply add print the value between steps.  So, in this case, you can add a line
```python
print(messaging)
```
between steps.

In [49]:
messaging = "Alice messaged Bob, but Bob didn't recognize Alice's new number."
messaging = messaging.replace("Bob", "Bobbie")
messaging = messaging.replace("Alice", "Bob")
messaging = messaging.replace("Bobbie", "Alice")
print (messaging)

Bob messaged Alice, but Alice didn't recognize Bob's new number.


---

Other string methods do not take any arguments at all, because the original string is all that's needed to compute the result. In this case, parentheses are still needed, but there's nothing in between the parentheses. Here are some methods that work that way:

|Method name|Value|
|-|-|
|`lower`|a lowercased version of the string|
|`upper`|an uppercased version of the string|
|`capitalize`|a version with the first letter capitalized|
|`title`|a version with the first letter of every word capitalized||


In [28]:
'unIverSITy of tennESSee'.title()

'University Of Tennessee'

All these string methods are useful, but most programmers don't memorize their names or how to use them.  In the "real world," people usually just search the internet for documentation and examples. A complete [list of string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) appears in the Python language documentation. [Stack Overflow](http://stackoverflow.com) has a huge database of answered questions that often demonstrate how to use these methods to achieve various ends.

### 2.2. Converting to and from Strings

Strings and numbers are different *types* of values, even when a string contains the digits of a number. For example, evaluating the following cell causes an error because an integer cannot be added to a string.

In [29]:
8 + "8"

TypeError: unsupported operand type(s) for +: 'int' and 'str'

However, there are built-in functions to convert numbers to strings and strings to numbers. 

    int:   Converts a string of digits to an integer ("int") value
    float: Converts a string of digits, perhaps with a decimal point, to a decimal ("float") value
    str:   Converts any value to a string

Try to predict what the following cell will evaluate to, then evaluate it.

In [ ]:
8 + int("8")

Suppose you're writing a program that looks for dates in a text, and you want your program to find the amount of time that elapsed between two years it has identified.  It doesn't make sense to subtract two texts, but you can first convert the text containing the years into numbers.

---

**Question 2.2.1:** Finish the code below to compute the number of years that elapsed between `one_year` and `another_year`.  Don't just write the numbers `1618` and `1648` (or `30`); use a conversion function to turn the given text data into numbers.

In [53]:
# Some text data:
one_year = "1618"
another_year = "1648"

# Complete the next line.  Note that we can't just write:
#   another_year - one_year
# If you don't see why, try seeing what happens when you
# write that here.
difference = int(another_year)-int(one_year)
difference

30

### 2.3. Strings as function arguments

String values, like numbers, can be arguments to functions and can be returned by functions.  The function `len` takes a single string as its argument and returns the number of characters in the string: its **len**-gth.  

Note that it doesn't count *words*. `len("We are the Tennessee Volunteers")` is 31, not 5.

---

**Question 2.3.1:**  Use `len` to find out the number of characters in the very long string in the next cell.  (It's the first sentence of the English translation of the 1789 French [Declaration of Rights](http://avalon.law.yale.edu/18th_century/rightsof.asp).)  The length of a string is the total number of characters in it, including things like spaces and punctuation.  Assign `sentence_length` to that number.

In [11]:
a_very_long_sentence = "The representatives of the French people, organized as a National Assembly, believing that the ignorance, neglect, or contempt of the rights of man are the sole cause of public calamities and of the corruption of governments, have determined to set forth in a solemn declaration the natural, unalienable, and sacred rights of man, in order that this declaration, being constantly before all the members of the Social body, shall remind them continually of their rights and duties; in order that the acts of the legislative power, as well as those of the executive power, may be compared at any moment with the objects and purposes of all political institutions and may thus be more respected, and, lastly, in order that the grievances of the citizens, based hereafter upon simple and incontestable principles, shall tend to the maintenance of the constitution and redound to the happiness of all."
sentence_length = len(a_very_long_sentence)
sentence_length

896

## 3. Importing code

Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Python allows us to **import** other code, creating a **module** that contains all of the names created by that code.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python. 

Suppose we want to very accurately compute the area of a circle with radius $5$ meters.  For that, we need the constant $\pi$, which is roughly $3.14$.  Conveniently, the `math` module has `pi` defined for us:

In [ ]:
import math
radius = 5
area_of_circle = radius**2 * math.pi
area_of_circle

`pi` is defined inside `math`, and the way that we access names that are inside modules is by writing the module's name, then a dot, then the name of the thing we want:

    <module name>.<name>
    
In order to use a module at all, we must first write the statement `import <module name>`.  That statement creates a module object with things like `pi` in it and then assigns the name `math` to that module.  Above we have done that for `math`.

---

**Question 3.1.** `math` also provides the name `e` for the base of the natural logarithm, which is roughly $2.71$.  Compute $e^{\pi}-\pi$, giving it the name `near_twenty`.

In [20]:
import math
near_twenty = math.e**math.pi-math.pi
near_twenty

19.99909997918947

### 3.1: Importing functions

**Modules** can provide other named things, including **functions**.  For example, `math` provides the name `sin` for the sine function.  Having imported `math` already, we can write `math.sin(3)` to compute the sine of $3$.  (Note that the input to this sine function is radians, not degrees; $180$ degrees is $\pi$ radians.)

---

**Question 3.1.1:** A $\frac{\pi}{4}$-radian ($45$-degree) angle forms a right triangle with equal base and height, pictured below.  If the hypotenuse (the radius of the circle in the picture) is $1$, then the height is $\sin\left(\frac{\pi}{4}\right)$.  Compute that using `sin` and `pi` from the `math` module.  Give the result the name `sine_of_pi_over_four`.

 <img src="trig.svg" width="300">

In [6]:
import math
sine_of_pi_over_four = math.sin(math.pi / 4)
sine_of_pi_over_four

0.7071067811865475

---

For your reference, here are some more examples of functions from the `math` module.

Note how different methods take in different number of arguments. Often, the documentation of the module will provide information on how many arguments is required for each method.

In [ ]:
# Calculating factorials.
math.factorial(5)

In [ ]:
# Calculating logarithms (the logarithm of 8 in base 2).
# The result is 3 because 2 to the power of 3 is 8.
math.log(8, 2)

In [ ]:
# Calculating square roots.
math.sqrt(5)

In [7]:
# Calculating cosines.
math.cos(math.pi)

-1.0

## 4. Arrays

Up to now, we haven't done much that you couldn't do yourself by hand, without going through the trouble of learning Python.  Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that's the result of multiplying each number in `billions_of_numbers` by .18 (18%).  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**, like a column in an Excel spreadsheet. 

<img src="excel_array.jpg">

### 4.1. Making arrays
You can type in the data that goes in an array yourself, but normally we create arrays by loading them from an external source, like a data file.

Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie").  It is standard to rename `numpy` as `np` for brevity:

```python
import numpy as np
```

In [ ]:
import numpy as np

Now, to create an array, call the function `np.array`. Note the brackts inside the function call, we use a list of arguments (the numbers in this case) to create a 1D array. Run this cell to see an example:

In [9]:
np.array([0.125, 4.75, -1.3])

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers $0.125$, $4.75$, and $-1.3$) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them names or use them as arguments to functions.

---

**Question 4.1.1:** Make an array containing the numbers 1, 2, and 3, in that order.  Name it `small_numbers`.

In [8]:
import numpy as np
small_numbers = np.array([1, 2, 3])
small_numbers

array([1, 2, 3])

**Question 4.1.2:** Make an array containing the numbers $0$, $1$, $-1$, $\pi$, and $e$, in that order.  Name it `interesting_numbers`.  *Hint:* How did you get the values $\pi$ and $e$ earlier?  You can refer to them in exactly the same way here.

**Note:** NumPy also provides the numbers $\pi$ (as `np.pi`) and $e$ (as `np.e`).  In general, if import `numpy`, you do not need the `math` module.

In [13]:
import numpy as np
import math
interesting_numbers = np.array([0,1,-1,math.pi,math.e])
interesting_numbers

array([ 0.        ,  1.        , -1.        ,  3.14159265,  2.71828183])

---

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  `np.arange(start, stop, space)` produces an array with all the numbers starting at `start` and counting up by `space`, stopping **before** `stop` is reached.  (*Be careful with this!*  It is a common source of errors to assume that `stop` might be included in the output!)

For example, the value of `np.arange(1, 6, 2)` is an array with elements 1, 3, and 5 -- it starts at 1 and counts up by 2, then stops **before** 6.  In other words, it's equivalent to `np.array([1, 3, 5])`.

`np.arange(4, 9, 1)` is an array with elements 4, 5, 6, 7, and 8.  (It doesn't contain 9 because `np.arange` stops *before* the stop value is reached.)

---

**Question 4.1.3:** Import `numpy` as `np` and then use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)

In [25]:
import numpy as np
multiples_of_99 = np.arange(0, 10000, 99)
multiples_of_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

#### Temperature readings
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the Blount County TN site for the month of December 2021.  To analyze the data, we want to know when each reading was taken, but we find that the data do not include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of December 2021 (midnight on December 1st) and each subsequent reading was taken exactly 1 hour after the last.

---

**Question 4.1.4:** Create an array of the *time, in seconds, since the start of the month* at which each hourly reading was taken.  Name it `collection_times`.

*Hint 1:* There were 31 days in December, which is equivalent to ($31 \times 24$) hours or ($31 \times 24 \times 60 \times 60$) seconds.  So your array should have $31 \times 24$ elements in it.

*Hint 2:* The `len` function works on arrays, too.  Check its length and make sure it has $31 \times 24$ elements.

In [35]:
import numpy as np
collection_times = np.arange(0,2678400,3600)
collection_times

array([      0,    3600,    7200,   10800,   14400,   18000,   21600,
         25200,   28800,   32400,   36000,   39600,   43200,   46800,
         50400,   54000,   57600,   61200,   64800,   68400,   72000,
         75600,   79200,   82800,   86400,   90000,   93600,   97200,
        100800,  104400,  108000,  111600,  115200,  118800,  122400,
        126000,  129600,  133200,  136800,  140400,  144000,  147600,
        151200,  154800,  158400,  162000,  165600,  169200,  172800,
        176400,  180000,  183600,  187200,  190800,  194400,  198000,
        201600,  205200,  208800,  212400,  216000,  219600,  223200,
        226800,  230400,  234000,  237600,  241200,  244800,  248400,
        252000,  255600,  259200,  262800,  266400,  270000,  273600,
        277200,  280800,  284400,  288000,  291600,  295200,  298800,
        302400,  306000,  309600,  313200,  316800,  320400,  324000,
        327600,  331200,  334800,  338400,  342000,  345600,  349200,
        352800,  356

*Optional check:*

In [18]:
# Not required
len(collection_times) == 31 * 24

True

If correct, you should get `True` above.  (But you might get `True` with an incorrect answer as well!)

### 4.2. Working with single elements of arrays ("indexing")
Let's work with a more interesting dataset.  The next cell creates an array called `population` that includes estimated world populations in every year from **1950** to roughly the present.  (The estimates come from the [United Nations](https://population.un.org/wpp/Download/Standard/Population/).)

Rather than type in the data manually, you should have downloaded a file called `world_population.csv`. 

In [19]:
# We use the package pandas to import the csv file.
import pandas as pd
data_file = pd.read_csv("world_population.csv")
# Quick look at the data:
data_file

Index  Position  Year  Population
0       0   Item #1  1950  2536431018
1       1   Item #2  1951  2584034227
2       2   Item #3  1952  2630861690
3       3   Item #4  1953  2677609061
4       4   Item #5  1954  2724846754
..    ...       ...   ...         ...
66     66  Item #67  2016  7464021934
67     67  Item #68  2017  7547858900
68     68  Item #69  2018  7631091113
69     69  Item #70  2019  7713468205
70     70  Item #71  2020  7794798729

[71 rows x 4 columns]

In [20]:
# Now choose one column of the data, 'Population':
population = data_file["Population"]
population

0     2536431018
1     2584034227
2     2630861690
3     2677609061
4     2724846754
         ...    
66    7464021934
67    7547858900
68    7631091113
69    7713468205
70    7794798729
Name: Population, Length: 71, dtype: int64

Here's how we get the first element of `population`, which is the world population in the first year in the dataset, 1950.

In [21]:
population[0]

2536431018

The value of that expression is the number 2536431018 (around 2.5 billion), because that's the first thing in the array `population`.

Notice that we used `population[0]`, not `population[1]`, to get the first element.  This is by convention; 0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `population`.  Read and run each cell.

In [22]:
# The third element in the array is the population
# in 1952.
population_1952 = population[2]
population_1952

2630861690

In [23]:
# The thirteenth element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population[12]
population_1962

3150420761

In [24]:
# The 71st element is the population in 2020.
population_2020 = population[70]
population_2020

7794798729

In [25]:
# The array has only 71 elements, so this doesn't work.
# (There's no element with 71 other elements before it.)
population_2021 = population[71]
population_2021

KeyError: 71

---

**Question 4.2.1:** Set `population_1973` to the world population in 1973, by getting the appropriate element from `population`.

In [26]:
population_1973 = population[23]
population_1973

3927780519

### 4.3. Doing something to every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to work with single elements.

#### Logarithms
Here is one simple question we might ask about world population:

> How big was the population in *orders of magnitude* in each year?

The logarithm function is one way of measuring how big a number is. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We could try using the `log10` function from the `NumPy` module, one element at a time:

In [27]:
import numpy as np
population_1950_magnitude = np.log10(population[0])
population_1951_magnitude = np.log10(population[1])
population_1952_magnitude = np.log10(population[2])
# etc
logged_populations = np.array([population_1950_magnitude, population_1951_magnitude])
logged_populations

array([9.40422306, 9.41229826])

But this is inefficient. Instead, we can ust use `np.log10` to takes the logarithm of each element of an array.  It takes a single array of numbers as its argument.  It returns an array of the same length, where the first element of the result is the logarithm of the first element of the argument, and so on.

---

**Question 4.3.1:** Use `np.log10` to compute the logarithms of the world population in every year.  Give the result (an array of 71 numbers) the name `population_magnitudes`.  Your code should be very short.

In [29]:
import numpy as np
population_magnitudes = np.log10(population)
population_magnitudes

0     9.404223
1     9.412298
2     9.420098
3     9.427747
4     9.435342
        ...   
66    9.872973
67    9.877824
68    9.882587
69    9.887250
70    9.891805
Name: Population, Length: 71, dtype: float64

---

<img src="array_logarithm.jpg">

This is called *elementwise* application of the function, since it operates separately on each element of the array it's called on.  The textbook's section on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.log10`.

#### Arithmetic
Arithmetic also works elementwise on arrays.  For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [ ]:
population_in_billions = ...
population_in_billions

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a tip (of $20\%$) on several restaurant bills at once (in this case just 3):

In [30]:
restaurant_bills = np.array([20.12, 39.90, 31.01])
print("Restaurant bills:\t", restaurant_bills)
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [20.12 39.9  31.01]
Tips:			 [4.024 7.98  6.202]


<img src="array_multiplication.jpg">

---

**Question 4.3.2:** Suppose the total charge at a restaurant is the original bill plus the tip.  That means we can multiply the original bill by $1.2$ to get the total charge.  Compute the total charge for each bill in `restaurant_bills`.

In [32]:
restaurant_bills = np.array([20.12, 39.90, 31.01])
total_charges = 1.2*restaurant_bills
total_charges

array([24.144, 47.88 , 37.212])

**Question 4.3.3:** Using `pd.read_csv`, open `more_restaurant_bills.csv`, which contains 100,000 bills in the column `'Bill'`.  Compute the total charge for each one.

In [39]:
more_restaurant_bills = pd.read_csv("more_restaurant_bills.csv")
extra_bills = more_restaurant_bills["Bill"]
more_total_charges = 1.2 * extra_bills
more_total_charges

0        20.244
1        20.892
2        12.216
3        11.808
4        22.128
          ...  
99995    35.736
99996     7.440
99997    19.308
99998    18.336
99999    35.664
Name: Bill, Length: 100000, dtype: float64

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 4.3.4:** What was the sum of all the bills in `more_restaurant_bills`, *including tips* (of $20\%$)?

In [43]:
import numpy as np
sum_of_bills = np.sum(more_restaurant_bills*1.2)
sum_of_bills

Bill    1795730.064
dtype: float64

**Question 4.3.5:** The powers of $2$ (i.e., $2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  Use `np.arange` and the exponentiation operator `**` to compute the first 30 powers of 2, starting from `2^0`.

In [48]:
import numpy as np
powers_of_2 = 2 ** np.arange(0, 30, 1)
powers_of_2

array([        1,         2,         4,         8,        16,        32,
              64,       128,       256,       512,      1024,      2048,
            4096,      8192,     16384,     32768,     65536,    131072,
          262144,    524288,   1048576,   2097152,   4194304,   8388608,
        16777216,  33554432,  67108864, 134217728, 268435456, 536870912])

## 5. Submission

To submit this lab, please download your notebook as a `.ipynb` file and submit in Canvas under *Lab 2* (under *Assignments*). To download, go to the toolbar at the top of this page, click **File** > **Download**. Then, go to our class's Canvas page and upload your file under "Lab 2". 

For easy identification, **please add your surname to the file**, as in: `lab_02_DS201_LastName.ipynb`.